#### Probem 5:

2520 is the smallest number that can be divided by each of the numbers from 1 to 10 without any remainder.

What is the smallest positive number that is *evenly divisible* by all of the numbers from 1 to 20?

*Evenly Divisible*: Adjective. Leaving no remainder when divided by. 15 is evenly divisible by 3, but 16 isn't.

In [1]:
# Find the smallest multiple by finding the Lowest Common Multiple for the set of numbers.
# To do this we need the prime factorisation of each number, then we take the highest power
# of each prime from the numbers and finally multiply them together.

In [2]:
# First we need to find the prime factorisation
# We have this from Problem 3:

# Function to find prime factors of a number
def prime_factors(number, prime_factor_list):
    # Will search numbers until a factor is found
    # Then divide number by our factor and search again
    # Return list of factors when number is 1
    if number == 1:
        return prime_factor_list
    for factor in range(2, number+1):
        if number % factor == 0:
            prime_factor_list = prime_factors(int(number/factor), prime_factor_list)
            prime_factor_list.append(factor)
            return(prime_factor_list)

In [3]:
# Now we get the prime factors for all the numbers up to n_max
def get_all_prime_factors(n_max):
    all_prime_factors = []
    for n in range(2,n_max+1):
        all_prime_factors.append(prime_factors(n, []))
    return(all_prime_factors)

In [4]:
# Now find all the unique primes to seach through
# Can use sets to reduce the duplicate primes
# Need to turn our prime factors list of lists into a flat list
def get_prime_factors_unique_list(all_prime_factors):
    prime_factors_flat_list = [prime_factor_value for prime_factors_list in all_prime_factors for prime_factor_value in prime_factors_list]
    prime_factors_unique_list = list(set(prime_factors_flat_list))
    return(prime_factors_unique_list)

In [5]:
# Now find highest number of occurances of each prime factor
def get_prime_factor_occurances(prime_factors_unique_list, all_prime_factors):
    number_occurances = []
    for n in prime_factors_unique_list:
        highest_occurance = 0
        for prime_factors_list_count in all_prime_factors:
            occurance = prime_factors_list_count.count(n)
            if occurance > highest_occurance: highest_occurance = occurance
        number_occurances.append(highest_occurance)
    return(number_occurances)

In [6]:
# Now take each unique prime and raise it to the powers of the number of occurances
def calculate_lowest_common_multiple(prime_factors_unique_list, number_occurances):
    lcm = 1
    for unique_prime in prime_factors_unique_list:
        lcm *= unique_prime ** number_occurances[prime_factors_unique_list.index(unique_prime)]
    return(lcm)

In [7]:
# Bring everything together
def find_solution(n_max):
    all_prime_factors = get_all_prime_factors(n_max)

    prime_factors_unique_list = get_prime_factors_unique_list(all_prime_factors)

    number_occurances = get_prime_factor_occurances(prime_factors_unique_list, all_prime_factors)
    
    solution = calculate_lowest_common_multiple(prime_factors_unique_list, number_occurances)
    
    solution_string = f'The Lowest Common Multiple of all integers from 1 to {n_max} = {solution:,} = '

    solution_string += ' x '.join(f'{unique_prime}^{number_occurances[prime_factors_unique_list.index(unique_prime)]}' for unique_prime in prime_factors_unique_list)

    print(solution_string)

In [8]:
find_solution(20)

The Lowest Common Multiple of all integers from 1 to 20 = 232,792,560 = 2^4 x 3^2 x 5^1 x 7^1 x 11^1 x 13^1 x 17^1 x 19^1


Problem 5 Solution: The Lowest Common Multiple of all integers from 1 to 20 = 232,792,560 = 2^4 x 3^2 x 5^1 x 7^1 x 11^1 x 13^1 x 17^1 x 19^1